src/context/TodoContext.tsx

```diff
@@ -0,0 +1,213 @@
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  const changeErrorMessage = useCallback((message: string) => {
+    setError(message);
+    setTimeout(() => {
+      setError('');
+    }, 3000);
+  }, []);
+
+  const focusOnInput = () => {
+    if (inputRef.current) {
+      inputRef.current.focus();
+    }
+  };
+
+  const clearInput = () => {
+    if (inputRef.current) {
+      inputRef.current.value = '';
+    }
+  };
+
+  useEffect(() => {
+    getTodos(USER_ID)
+      .then(data => {
+        setTodos(data);
+        focusOnInput();
+      })
+      .catch(() => changeErrorMessage(ErrorMessages.DOWNLOAD));
+  }, [changeErrorMessage]);
+
+  const filteredTodos = useMemo(() => {
+    switch (filter) {
+      case 'active':
+        return todos.filter(todo => !todo.completed);
+
+      case 'completed':
+        return todos.filter(todo => todo.completed);
+
+      case 'all':
+      default:
+        return todos;
+    }
+  }, [filter, todos]);
+
+  const completedTodosId = useMemo(() => {
+    return todos.filter(
+      todo => todo.completed,
+    ).reduce<number[]>((acc, next) => {
+      return [...acc, next.id];
+    }, []);
+  }, [todos]);
+
+  const addTodo = useCallback((todo: string) => {
+    const newTodo = {
+      id: 0,
+      title: todo,
+      userId: USER_ID,
+      completed: false,
+    };
+
+    setTempTodo(newTodo);
+
+    createTodo(newTodo)
+      .then(data => {
+        setTodos(old => old.concat(data));
+        clearInput();
+        focusOnInput();
+      })
+      .catch(() => changeErrorMessage(ErrorMessages.ADD))
+      .finally(() => {
+        setTempTodo(null);
+      });
+  }, [changeErrorMessage]);
+
+  const removeTodo = useCallback((itemId: number) => {
+    setLoadingTodoId(prev => [...prev, itemId]);
+    deleteTodo(itemId)
+      .then(() => {
+        setTodos(old => old.filter(todo => todo.id !== itemId));
+      })
+      .catch(() => changeErrorMessage(ErrorMessages.DELETE))
+      .finally(
+        () => {
+          setLoadingTodoId(prev => prev.filter(id => id !== itemId));
+          focusOnInput();
+        },
+
+      );
+  }, [changeErrorMessage]);
+
+  const updateTodo = (itemId: number, completed: boolean) => {
+    setLoadingTodoId(prev => [...prev, itemId]);
+    patchTodo(itemId, !completed)
+      .then(() => {
+        setTodos(old => old.map(todo => {
+          if (todo.id === itemId) {
+            return {
+              ...todo,
+              completed: !completed,
+            };
+          }
+
+          return todo;
+        }));
+        focusOnInput();
+      })
+      .catch(() => changeErrorMessage(ErrorMessages.UPDATE))
+      .finally(
+        () => setLoadingTodoId(prev => prev.filter(id => id !== itemId)),
+      );
+  };
+
+  const removeAllCompleted = useCallback(() => {
+    todos.forEach(todo => {
+      if (todo.completed) {
+        removeTodo(todo.id);
+      }
+    });
+  }, [todos, removeTodo]);
+
+  const nrOfActiveTodos = useMemo(() => {
```

Comment:
Do we really have to memoize it? 
It's just a simple `filter`: memorizing it may more expencive operation than filtering itself.